### PR TITLE
[RFC FS-1017] Allow all inferrable SRTP constraints to be written in signatures

### PR DESCRIPTION
**Pull request by dsyme**
This implements [RFC FS--1017](https://github.com/fsharp/FSharpLangDesign/blob/master/RFCs/FS-1017-fix-srtp-constraint-parsing.md). A niggling language issue [reported by @drvink quite a long time ago.](https://fslang.uservoice.com/forums/245727-f-language/suggestions/7887270-allow-all-inferrable-srtp-constraints-to-be-writte )  

Basically it allows statically resolved member constraints to include a type - rather than a type parameter - in the list of items by which the constraint can be solved.

```fsharp
     .. when (Ty or  ^Typar) : (static member ... :  ...) =
```

This both fixes a "hole" where soome types could be inferred and printed, but not written in signatures, and also increases the utility of the statically-resolved-type-parameter (SRTP) mechanism.

The implementation is very straight-forward.

### CODE REVIEW

**Comment by drvink** [comment]
RFC PR at fsharp/FSharpLangDesign#103.

File: src/fsharp/TypeChecker.fs [hunk 1]
```diff
@@ -4118,13 +4118,13 @@ let rec TcTyparConstraint ridx cenv newOk checkCxs occ (env: TcEnv) tpenv c =
...
-and TcPseudoMemberSpec cenv newOk env synTypars tpenv memSpfn m = 
+and TcPseudoMemberSpec cenv newOk env synTypes tpenv memSpfn m = 
 #if ALLOW_MEMBER_CONSTRAINTS_ON_MEASURES
     let tps,tpenv = List.mapFold (TcTyparOrMeasurePar None cenv env newOk) tpenv synTypars
 #else
-    let tps,tpenv = List.mapFold (TcTypar cenv env newOk) tpenv synTypars
+    let tys,tpenv = List.mapFold (TcTypeAndRecover cenv newOk CheckCxs ItemOccurence.UseInType env) tpenv synTypes
 #endif
-    let tys = List.map mkTyparTy tps
+    //let tys = List.map mkTyparTy tps
```
Comment (KevinRansom) [inline src/fsharp/TypeChecker.fs]: probably should be deleted.

**Comment by KevinRansom** [comment]
other than commented out code looks fine to me.

**Comment by dsyme** [comment]
I've updated this PR with a fix for https://github.com/Microsoft/visualfsharp/issues/1306

**Comment by dsyme** [comment]
Transient AppVeyor failure (unrelated to PR).  AppVeyor have changed their hardware recenty which may explain a change in test timings for non-deterministic tests.  There is either a bug lurking or an unstable test

```
[00:40:44] 1) Error : FSharp.Core.Unittests.FSharp_Core.Microsoft_FSharp_Control.AsyncModule.Async.Choice specification test
[00:40:44] System.Exception : Falsifiable, after 43 tests (0 shrinks) (StdGen (15424488,296180491)):
[00:40:44] Original:
[00:40:44] ChoiceWorkflow
[00:40:44]   ([NoneResultAfter -8; NoneResultAfter 9; SomeResultAfter 2;
[00:40:44]     SomeResultAfter -18; ExceptionAfter 12; ExceptionAfter -8;
[00:40:44]     NoneRes
[00:40:44] ultAfter -15; NoneResultAfter 2],Some 8)
[00:40:44] with exception:
[00:40:44] NUnit.Framework.AssertionException:   Expected: 4000
[00:40:44]   But was:  6000
```

**Pull request merged**